### PR TITLE
tools/trap: Refactor the trap tool to make the script easy to understand

### PR DESCRIPTION
**Pull request by neel-samsung**
tools/trap: Refactor the trap tool to make the script easy to understand
1. We refactor the trap script into multiple files. Earlier, there were only two files: ramdumpParser.py and debugsymbolviewer.py
2. We divided the script content into multiple files: trap.py, dumpParser.py, logParser.py, heapNode.py and utils.py
   - trap.py        : It is the wrapper script which contains usage API, initializing code and make instance of relevant class that provides APIs to parse the log file and ram dump
   - logParser.py   : It has logParser class which provide APIs (i.e. find crash point, call stack details, heap corruption, etc) to parse the assert/crash log file
   - dumpParser.py  : It has dumpParser class which has APIs mainly for parsing RAM/Flash dump
   - heapNode.py    : It contains dumpNode class which helps in parsing heap corruption log.
   - utils.py       : It act as utility for all above files and provide API for setting up symbol table and retrieving symbol based on address and vice versa.
3. Usage for TRAP tool: python3 trap.py -t <log-file> -c <config-path> -b <bin-path> -e <elf-path>
   Instead of ramdumpParser.py, we use trap.py as script.

### CODE REVIEW

**Review by sunghan-chang — COMMENTED** [review]
1. I don't get difference between assertlogParser and ramdumpParser. Could you let me know?
2. Is it ok with 644 for py file?

**Comment by sunghan-chang** [comment]
@neel-samsung And armv7-a assert file modification is not a part of "Refactor the trap tool". Please split the commit.

**Comment by neel-samsung** [comment]
> 1. I don't get difference between assertlogParser and ramdumpParser. Could you elaborate more?
> 
> ```
> assertlogParser.py : It contains the parsing logic to parse the assert/crash log file
> ramdumpParser.py : It will be mainly the helper script that contains a dumpParser class which has APIs mainly for parsing RAM/Flash dump
> ```
> 
> 2. Is it ok with permission 644 for py file?

1. Earlier, we used to support the debugging using ramdump file. Now, we don't support debugging using ramdump log for armv7-a and armv8-m. we only support trap script to work for assert/crash log.
 Now, ```ramdumpParser.py``` contains parsing logic mainly for this ramdump log and ```assertlogParser.py``` contains parsing logic  for assert/crash log.

2. We won't be executing the  ```ramdumpParser.py```. It has only ```dumpParser``` class. It will get imported only. So, we can keep 644.

**Comment by neel-samsung** [comment]
@r-prabu @kishore-sn Please Review it once.

File: tools/trap/assertlogParser.py [hunk 1]
```diff
@@ -0,0 +1,1192 @@
...
+import string
+import sys, time
+import struct
+import subprocess
+from getopt import GetoptError, getopt as GetOpt
+from array import *
+from ramdumpParser import dumpParser
+
+# Global variables
+FP = 11
```
Comment (r-prabu) [inline tools/trap/assertlogParser.py:35]: Will this fixed register numbers work for all processor arch?

**Comment by sunghan-chang** [comment]
@neel-samsung Hmmm. You split the single file to two files. But still they are much big files, over 1000 lines and 400 lines. It looks they are still difficult to analyze the script.
> Usage for TRAP tool: python assertlogParser.py -t
Instead of ramdumpParser.py, we use assertlogParser.py as it contains parsing logic.

Then, you should modify [README](https://github.com/Samsung/TizenRT/blob/master/tools/trap/HowToUseTrap.md).

**Comment by neel-samsung** [comment]
> @neel-samsung Hmmm. You split the single file to two files. But still they are much big files, over 1000 lines and 400 lines. It looks they are still difficult to analyze the script.
> 
> > Usage for TRAP tool: python assertlogParser.py -t
> > Instead of ramdumpParser.py, we use assertlogParser.py as it contains parsing logic.
> 
> Then, you should modify [README](https://github.com/Samsung/TizenRT/blob/master/tools/trap/HowToUseTrap.md).

yes, I agree. So, we split the TRAP tool into 5 different files and also most of the part of code was based on functional Programming, we changed it to OOP. 
   - trap.py        : It is the wrapper script which contains usage API, initializing code and make instance of relevant class that provides APIs to parse the log file and ram dump
   - logParser.py   : It has logParser class which provide APIs (i.e. find crash point, call stack details, heap corruption, etc) to parse the assert/crash log file
   - dumpParser.py  : It has dumpParser class which has APIs mainly for parsing RAM/Flash dump
   - heapNode.py    : It contains dumpNode class which helps in parsing heap corruption log.
   - utils.py       : It act as utility for all above files and provide API for setting up symbol table and retrieving symbol based on address and vice versa.

Please check and Review. Let us know your feedback.

**Comment by neel-samsung** [comment]
> > @neel-samsung Hmmm. You split the single file to two files. But still they are much big files, over 1000 lines and 400 lines. It looks they are still difficult to analyze the script.
> > > Usage for TRAP tool: python assertlogParser.py -t
> > > Instead of ramdumpParser.py, we use assertlogParser.py as it contains parsing logic.
> > 
> > 
> > Then, you should modify [README](https://github.com/Samsung/TizenRT/blob/master/tools/trap/HowToUseTrap.md).
> 
> yes, I agree. So, we split the TRAP tool into 5 different files and also most of the part of code was based on functional Programming, we changed it to OOP.
> 
> * trap.py        : It is the wrapper script which contains usage API, initializing code and make instance of relevant class that provides APIs to parse the log file and ram dump
> * logParser.py   : It has logParser class which provide APIs (i.e. find crash point, call stack details, heap corruption, etc) to parse the assert/crash log file
> * dumpParser.py  : It has dumpParser class which has APIs mainly for parsing RAM/Flash dump
> * heapNode.py    : It contains dumpNode class which helps in parsing heap corruption log.
> * utils.py       : It act as utility for all above files and provide API for setting up symbol table and retrieving symbol based on address and vice versa.
> 
> Please check and Review. Let us know your feedback.

@sunghan-chang @kishore-sn Please Review.

File: tools/trap/dumpParser.py [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+				low = mid + 1
+
+
+			mid = (high + low) >> 1
+
+		print("Sorry. Given address does not present in stack lookup")
+		return None
+
+	# Function to Parse the i/p log file (which contains stackdump during assert) passed as part of -t argument ( No dump is supplied )
+	def parse_log_file(self,log_file,debug=False):
```
Comment (kishore-sn) [inline tools/trap/dumpParser.py:140]: There is something wrong here. Parsing log file is only supposed to be done in log parser. Not in dump parser. Dump parser is supposed to be used only when log file is not given. Can you pls recheck this?

File: tools/trap/logParser.py [hunk 3]
```diff
@@ -0,0 +1,796 @@
...
+register_dump = "Asserted task's register dump\n"
+BIN_ADDR_FXN = "Loading location information\n"
+
+partition_string = "==========================================================="
+
+class logParser:
+
+	# Init function for logParser Class
+	def __init__(self, dump_file = None, elf = None, log_file = None, bin_path = None, config_path= None, xip_enabled = False, have_ram_kernel_text = False):
+		self.dump_file = dump_file		     # dump file
```
[the remark this answers is not in the document]
Comment (neel-samsung) [reply tools/trap/logParser.py:44]: It is required while printing TRAP header output.

File: tools/trap/dumpParser.py (hunk 2 above)
Comment (neel-samsung) [reply tools/trap/dumpParser.py:140]: So, Earlier dump parser was also responsible for parsing the log file. But now, for call stack details, both dump parser and log parser are parsing the log file. Is it feasible to remove this ```parse_log_file``` API as we are already parsing in log parser? But I don't know about earlier arch (armv7-r), whether they support log Parsing from ```logParser``` class.

File: tools/trap/trap.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+		lparser.format_log_file()
+
+		# Get the number of application binaries, names, text address and sizes
+		lparser.find_number_of_binaries()
+
+		# Print TRAP output Header information i.e. log_file, dump file, elf files, etc.
+		lparser.print_trap_output_header()
+		
+		# Calling the Constructor with the initial set of arguments
+		rParser = dumpParser(dump_file = dump_file, elf = elf, log_file = log_file, bin_path = BIN_PATH, config_path = CONFIG_PATH, have_ram_kernel_text = have_ram_kernel_text, debug=False)
```
[the remark this answers is not in the document]
Comment (neel-samsung) [reply tools/trap/trap.py:191]: Thank you for Reviewing.. I will take care of all comments.

**Review by sunghan-chang — COMMENTED** [review]
Need to check dutils.sh. It calls previous sh file for trap.

**Review by sunghan-chang — COMMENTED** [review]
@neel-samsung Sorry for no time to review all because it's huge.. When your team review is done, let's merge.
@r-prabu @kishore-sn Could you review and merge this?

**Comment by sunghan-chang** [comment]
> Need to check dutils.sh. It calls previous sh file for trap.

@neel-samsung Please confirm whether TRAP is working well through the dutils.sh or not.

**Comment by neel-samsung** [comment]
> > Need to check dutils.sh. It calls previous sh file for trap.
> 
> @neel-samsung Please confirm whether TRAP is working well through the dutils.sh or not.

yes.. I have checked and confirm it's working well through dutils.sh.

File: tools/trap/cli/trap.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+			# Function to parse the assert/crash log
+			lparser.parse_log()
+
+		# Get ARM arch family
+		elif arch_family:
+			#If architecture is cortex M or A, then run return without further script execution
+			usage()
+			return None
+		else:
```
[the remark this answers is not in the document]
Comment (neel-samsung) [reply tools/trap/cli/trap.py:224]: yes.. It will be better.

File: tools/trap/cli/logParser.py [hunk 6]
```diff
@@ -0,0 +1,779 @@
...
+
+		# print current running task List
+		self.print_task_list()
+
+	# Function to format logs and delete the timestamp (format-|xxxxxxxxx|) if it consists of timestamp at the start of each log line
+	def format_log_file(self):
+
+		# Delete unwanted logs (if any) and timestamp at the start of each line
+		with open(self.log_file, "r") as f:
+			data = f.readlines()
```
[the remark this answers is not in the document]
Comment (neel-samsung) [reply tools/trap/cli/logParser.py:717]: If we are opening file with 'with open', then it gets auto close when control leaves out of block.

**Review by kishore-sn — APPROVED** [review]

**Pull request merged**